compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.ir.asSimpleLambda
+import org.jetbrains.kotlin.backend.common.ir.inline
+import org.jetbrains.kotlin.backend.common.lower.*
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.config.JVMAssertionsMode
+import org.jetbrains.kotlin.ir.IrElement
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.buildField
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetObjectValueImpl
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPackageFragment
+import org.jetbrains.kotlin.ir.util.patchDeclarationParents
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformer
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+internal val assertionPhase = makeIrFilePhase(
+    ::AssertionLowering,
+    name = "Assertion",
+    description = "Lower assert calls depending on the assertions mode"
+)
+
+private class AssertionLowering(private val context: JvmBackendContext) :
+    FileLoweringPass,
+    IrElementTransformer<AssertionLowering.ClassInfo?>
+{
+    // Keeps track of the $assertionsDisabled field, which we generate lazily for classes containing
+    // assertions when compiled with -Xassertions=jvm.
+    private class ClassInfo(val irClass: IrClass, var assertionsDisabledField: IrField? = null)
+
+    override fun lower(irFile: IrFile) {
+        // In legacy mode we treat assertions as inline function calls
+        if (context.state.assertionsMode != JVMAssertionsMode.LEGACY)
+            irFile.transformChildren(this, null)
+    }
+
+    override fun visitClass(declaration: IrClass, data: ClassInfo?): IrStatement {
+        val info = ClassInfo(declaration)
+
+        super.visitClass(declaration, info)
+
+        // Note that it's necessary to add this member at the end of the class, behind all user-visible initializers,
```

Comment:
Could you point to specification?
Actually I see totatly opposite situation for Java code:
```
class A2 {

    static long z = System.currentTimeMillis();

    public static void main(String foo) {
        assert foo != null : "123";
    }
}
```

Reply:
It's a bit more subtle than I thought. The java spec says (se7, section 14.10):

> An assert statement that is executed before its class has completed initialization is enabled.

However, the idea here is that the assertion status is initialized as soon as the initializer runs. The rule is only relevant with circular initializers for which we run static code before running the initializer at all. I've changed the comment and code accordingly and expanded the corresponding test.